include/revng/ADT/UpcastablePointer.h

```diff
@@ -221,17 +204,28 @@ class UpcastablePointer {
       Other.upcast([&](auto &OtherUpcasted) {
         using ThisType = std::remove_cvref_t<decltype(Upcasted)>;
         using OtherType = std::remove_cvref_t<decltype(OtherUpcasted)>;
-        if constexpr (std::is_same_v<ThisType, OtherType>) {
+        if constexpr (std::is_same_v<ThisType, OtherType>)
           Result = Upcasted == OtherUpcasted;
-        }
       });
     });
     return Result;
   }
 
-  auto get() const noexcept { return Pointer.get(); }
-  auto &operator*() const { return *Pointer; }
-  auto *operator->() const noexcept { return Pointer.operator->(); }
+  template<UpcastablePointerLike PointerLike>
+    requires(std::three_way_comparable_with<T, Dereferenced<PointerLike>>)
+  auto operator<=>(const PointerLike &Other) const {
+    return *Pointer <=> *Other;
+  }
```

Comment:
This implementation is significantly different from `operator==`, how comes?

Reply:
Mostly because it requires the `Kind()` and I didn't feel comfortable mentioning it in here.

The idea of this is simple: if underlying type is `three_way_comparable_with`, just path this call through to it, otherwise fail the compilation.